platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+the
+network's nodes. This dynamic list of Addresses will be continuously updated by HAPI transactions, such as those that
+create, update, or delete nodes.
+At some point when the App decides it's time to begin work on adopting a new address book (scope beyond this proposal),
+it will create a `Roster` object with information from the Address Book and pass it to the platform.
+
+The mechanism for doing so is detailed below.
+
+## What this proposal does not address
+
+## Requirements
+
+- Immutability: The roster hash will be a critical piece of data. Rosters must be immutable to protect the integrity
+  of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the App to create and submit candidate rosters to the platform,
+  and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use the new `Roster` object instead.
+- Address book paradigm discontinued within the platform codebase in favor of `Roster`s
+
+## Architecture
+
+###### Roster Lifecycle
+
+![](TSS%20Roster%20Lifecycle.drawio.svg)
+
+### Roster Public API
+
+#### Setting the Candidate Roster
+
+A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+
+```java
+//in PlaformState
+
+void setCandidateRoster(@NonNull final Roster candidateRoster);
+```
+
+The steps to validate and add the Roster to the state will be executed within the same thread that submits the request.
+An exception will be thrown if the Roster is invalid.
+The durability of the submitted Roster is ensured via storing it in the state.
+
+The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.
+However, its replacement will continue to have this new API to set a candidate roster.
+
+## Data Structure
+
+The App is already responsible for managing the address book.
+We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object
+and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.
+
+### Protobuf
+
+#### Roster
+
+The protobuf definition for the Roster will look as follows:
+
+```proto
+/**
+ * A single roster in the network state.
+ * <p>
+ * The roster SHALL be a map of key `NodeId`, value `RosterEntry`.
+ */
+message Roster {
+
+    /**
+     * Map of roster entries, one per consensus node.
+     * <p>
+     * This map SHALL contain roster entries in natural order of ascending node ids.
+     * The node Id key SHALL match the node Id in the value.
+     * This map SHALL NOT be empty.<br/>
+     */
+    map<uint64, RosterEntry> rosterEntries = 1;
```

Comment:
I would suggest
```
repeated RosterEntry roster_entries = 1;
```
because serialization will be faster, the serialized size will be smaller, and the overhead of building a `Map<Long, RosterEntry>` from the list will be vanishingly small.

Reply:
The original proposal was a list but changed to a map after a feedback that we do need the roster entries in a form where their keys are sorted.

If i understand you correctly, you're saying we should still build and hold a map in memory, but store a list (of the map's entries) in the state?

Reply:
While I don't have a strong opinion on this, I'll just note that a nice property of the protobuf map is the deterministic serialization of the map object. With a list, the serialized bytes will depend on the order of items in the list, making the user of the model responsible for deterministic binary representation. With the map, the user doesn't need to worry about that and they always get the deterministic result - all entries are always serialized in the natural order of the keys.

The determinism of the binary form might be important for hashing/signing and similar use-cases.

Reply:
If our intent is to use the protobuf as in-memory storage instead of copying the data into a Java Map on deserialization, then lookup on a list is O(n), but O(1) in a hash table and O(logn) in a tree.  I don't know what the backing implementation is for the State API, but it has to be better than lookup in a list.

Reply:
It probably does not matter here, but the asymptotic growth rate (big o notation) is irrelevant for small datasets. Lookups in small ArrayLists are usually faster than in HashMaps.

Reply:
https://protobuf.dev/programming-guides/proto3/#maps seems to say that maps are dangerous from a deterministic wire format perspective?

> Wire format ordering and map iteration ordering of map values are undefined, so you cannot rely on your map items being in a particular order.

Reply:
In PBJ, we made map serialization deterministic though. We always serialize them in the natural order of keys. And upon deserialization, we always store them in PbjMap objects which expose the order via their `public List<K> getSortedKeys()` method. I should've clarified that in my comment above instead of calling them "protobuf maps".

So unless we want to stop using PBJ, this shouldn't be an issue.

Reply:
It may be important to note that these values (and all values in state) will be part of the _public API_ in the Block Stream, and may be processed by protoC-generated code (or other protobuf compilers for languages, like Rust, that protoC doesn't support).

Those other systems may _not_ be deterministic for map ordering (possibly making it very difficult for those systems to correctly recreate state).

Just something to consider.

Reply:
We decided to go with the recommendation from Michael and forbid the use of maps in state. While it would be awesome to have it be a map, so that the generated pbj classes would give us a map that behaved just as we liked, the fact is that map on-wire serialization is non-deterministic across multiple protoc implementations, and since state is intended to be parsed and used by downstream applications across multiple languages, using map will create friction for those people.